### DISCUSSION TITLE
Is there a reason why the read/write lock on Windows is a CRITICAL_SECITON+Semaphore instead of just SRWLock?

### DISCUSSION THREAD

**Discussion by dmachaj** [libuv/libuv#3378] [Q&A] [answered]
I happened to be reading through some of the threading code for Windows to satisfy my own curiosity and I noticed that libuv creates its own read/write lock out of a critical section and a semaphore.  This seems curious to me because Windows has an OS primitive that can solve this same problem.  The OS primitive has the benefit that it is quite cheap compared to a critical section; it is the size of a pointer.  The biggest risk that I'm aware of is that the lock cannot be safely acquired _recursively_ in shared mode.

See https://docs.microsoft.com/en-us/windows/win32/sync/slim-reader-writer--srw--locks for the documentation on this lock type.

I'm curious whether there is a reason why the OS primitive is not being used, such as recursive shared acquire, or if this is a good opportunity for a performance improvement.  Thanks.

(Note: I am a Microsoft employee)

**Comment by bnoordhuis** [comment, marked as answer]
It's a relic from when libuv supported Windows XP.  We no longer do so if you want to send a pull request, that'd be very welcome.

**Reply by dmachaj** [reply]
That's exactly the kind of answer I was hoping for 👍.  I'm interested in sending a PR and will try to have something next week.

**Reply by dmachaj** [reply]
Issue: https://github.com/libuv/libuv/issues/3382
PR: https://github.com/libuv/libuv/pull/3383
